lua/quicktest/adapters/criterion/init.lua

```diff
@@ -0,0 +1,166 @@
+local Job = require("plenary.job")
+
+local util = require("quicktest.adapters.criterion.util")
+local meson = require("quicktest.adapters.criterion.meson")
+local criterion = require("quicktest.adapters.criterion.criterion")
+
+local ns = vim.api.nvim_create_namespace("quicktest-criterion")
+
+local M = {
+  name = "criterion",
+  test_results = {},
+  builddir = "build",
+}
+
+---@class Test
+---@field test_suite string
+---@field test_name string
+
+---@class CriterionTestParams
+---@field test Test
+---@field test_exe string
+---@field bufnr integer
+---@field cursor_pos integer[]
+
+---@param bufnr integer
+---@param cursor_pos integer[]
+---@return CriterionTestParams
+M.build_file_run_params = function(bufnr, cursor_pos)
+  local test_exe = util.get_test_exe_from_buffer(bufnr, M.builddir)
+
+  if test_exe == "" then
+    return {}
+  end
+
+  return {
+    test = {},
+    test_exe = test_exe,
+    bufnr = bufnr,
+    cursor_pos = cursor_pos,
+  }
+end
+
+---@param bufnr integer
+---@param cursor_pos integer[]
+---@return CriterionTestParams
+M.build_line_run_params = function(bufnr, cursor_pos)
+  local test_exe = util.get_test_exe_from_buffer(bufnr, M.builddir)
+  local line = criterion.get_nearest_test(bufnr, cursor_pos)
+
+  if line == "" or test_exe == "" then
+    return {}
+  end
+
+  local test = criterion.get_test_suite_and_name(line)
+  return {
+    test = test,
+    test_exe = test_exe,
+    bufnr = bufnr,
+    cursor_pos = cursor_pos,
+  }
+end
+
+--- Determines if the test can be run with the given parameters.
+--- Attempt to run if params is not an empty table
+---@param params CriterionTestParams
+---@return boolean
+M.can_run = function(params)
```

Comment:
Not needed any more, cause build_* functions return error as second return value